src/libfetchers/input-accessor.cc

```diff
@@ -61,11 +61,27 @@ SourcePath SourcePath::parent() const
     return {accessor, std::move(*p)};
 }
 
+SourcePath SourcePath::followSymlinks() const {
+    SourcePath path = *this;
+    unsigned int followCount = 0, maxFollow = 1000;
+
+    /* If `path' is a symlink, follow it.  This is so that relative
+       path references work. */
+    while (true) {
+        // Basic cycle/depth limit to avoid infinite loops.
+        if (++followCount >= maxFollow)
+            throw Error("too many levels of symbolic links while traversing the path '%s'; assuming it leads to a cycle after following %d indirections", this->to_string(), maxFollow);
```

Comment:
```suggestion
            throw Error("cycle detected following symlink '%s'", this->to_string());
```
After 1000 links we can safely assume there's a cycle, so no need to be very verbose.